File: 1-100/51-n-queens.cpp
vector<vector<string>> solveNQueens(int n) {
    vector<vector<string>> ans;
    if (n == 0) 
        return ans;
    vector<string> borad(n, string(n, '.'));
    vector<bool> column(n, false), ldiag(2*n-1, false), rdiag(2*n-1, false);
    backtracking(borad, column, ldiag, rdiag, ans, 0, n);
    return ans;
}

void backtracking(vector<string> &borad, vector<bool> &column, vector<bool> &ldiag,
    vector<bool> &rdiag, vector<vector<string>> &ans, int row, int n) {
    if (row == n) {
        ans.push_back(borad);
        return;
    }
    for (int col = 0; col < n; ++col) {
        if (column[col] || ldiag[n-row-1+col] || rdiag[col+row])
            continue;
        borad[row][col] = 'Q';
        column[col] = ldiag[n-row-1+col] = rdiag[col+row] = true;
        backtracking(borad, column, ldiag, rdiag, ans, row + 1, n);
        column[col] = ldiag[n-row-1+col] = rdiag[col+row] = false;
        borad[row][col] = '.';
    }
}

File: python version/51-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        cols = [False] * n
        l_diags = [False] * (2*n-1)
        r_diags = [False] * (2*n-1)
        board = [["."]*n for _ in range(n)]
        ans = []
        def dfs(row):
            if row == n:
                tmp = ["".join(i) for i in board]
                ans.append(tmp)
                return
            for col in range(n):
                if cols[col] or l_diags[row+col] or r_diags[n-1-col+row]: continue
                cols[col] = l_diags[row+col] = r_diags[n-1-col+row] = True
                board[row][col] = "Q"
                dfs(row+1)
                board[row][col] = "."
                cols[col] = l_diags[row+col] = r_diags[n-1-col+row] = False
        dfs(0)
        return ans

